scheduler/cloudformation/cloudformation.go

```diff
@@ -280,13 +288,111 @@ func (s *Scheduler) submit(ctx context.Context, tx *sql.Tx, app *scheduler.App,
+		// FIXME: At this point, we don't want to affect UX by waiting
+		// around, so we return. But, if the stack update times out, or
+		// there's an error, that information is essentially silenced.
+		return nil
+	case err = <-errCh:
+	case <-locked:
+		// if a lock is obtained within the time frame, we might as well
+		// just wait for the update to get submitted.
+		err = <-errCh
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// If wait is set, we'll wait for the stack update to fully complete.
+	if wait {
+		return <-done
+	}
+
+	return nil
+}
+
+// enqueueStackUpdate enqueues the stack update. This function returns as soon
+// as the stack update has been submitted.
+//
+// * If there are no updates currently in progress, the stack will be updated .
+// * If there is a currently active stack update, the update will be queued behind it.
 //
-// TODO: Timeout?
-func (s *Scheduler) updateStack(stack *cloudformation.Stack, input *cloudformation.UpdateStackInput, wait bool) (*cloudformation.UpdateStackOutput, error) {
-	status := *stack.StackStatus
-	stackName := input.StackName
+// An error will be sent on the `done` channel when the stack update completes.
+func (s *Scheduler) enqueueStackUpdate(input *cloudformation.UpdateStackInput, locked chan struct{}, done chan error) error {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	timeout := int(lockTimeout.Seconds() * 1000)
+	_, err = tx.Exec(fmt.Sprintf("SET LOCAL lock_timeout = %d", timeout))
+	if err != nil {
+		return fmt.Errorf("error setting lock timeout: %v", err)
+	}
+
+	// We have one lock per CloudFormation stack.
+	key := crc32.ChecksumIEEE([]byte(fmt.Sprintf("stack_%s", *input.StackName)))
+
+	// If there are any other pending updates for this stack, fuck em.
+	_, err = tx.Exec(`SELECT pg_cancel_backend(pending.pid) FROM (SELECT pid FROM pg_locks WHERE locktype = 'advisory' AND granted = 'f' AND objid = $1) as pending`, key)
+	if err != nil {
+		return fmt.Errorf("error canceling pending locks: %v", err)
+	}
+
+	context := fmt.Sprintf("stack %s", *input.StackName)
+	if input.TemplateURL != nil {
+		context = fmt.Sprintf("%s (%s)", context, *input.TemplateURL)
+	}
+
+	_, err = tx.Exec(fmt.Sprintf("SELECT pg_advisory_lock($1) /* %s */", context), key)
+	if err != nil {
+		// TODO: If this was a user cancelation, ignore.
+		return fmt.Errorf("error obtaining lock to update stack %s: %v", *input.StackName, err)
+	}
+
+	close(locked)
+
+	// Once the lock has been obtained, let's submit the stack update.
+	err = s.executeStackUpdate(input)
+	if err != nil {
+		return err
+	}
+
+	// Start up a goroutine that will wait for this stack update to
+	// complete, and release the lock when it completes.
+	go s.waitUntilStackUpdateComplete(tx, *input.StackName, done)
+
+	return nil
+}
+
+func (s *Scheduler) waitUntilStackUpdateComplete(tx *sql.Tx, stackName string, done chan error) {
+	err := s.cloudformation.WaitUntilStackUpdateComplete(&cloudformation.DescribeStacksInput{
```

Comment:
This should have some reasonable timeout (like 10 minutes or so), just to ensure that this can't hold the lock forever.